### PR TITLE
improvement: count() now costs O(1) instad of O(n)

### PR DESCRIPTION
**Pull request by snipershady**
Now the method `count()` costs O(1) instad of O(n)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,5 @@
 /phpcs.xml
 /.phpcs-cache
 /.phpunit.result.cache
+/nbproject/private/
+/nbproject/
```
Comment (derrabus) [inline .gitignore]: Please revert.

File: src/ArrayCollection.php [hunk 2]
```diff
@@ -65,6 +71,7 @@ class ArrayCollection implements Collection, Selectable, Stringable
     public function __construct(array $elements = [])
     {
         $this->elements = $elements;
+        $this->size = count($elements);
```
Comment (derrabus) [inline src/ArrayCollection.php]: This increases the cost of the constructor call from O(1) to O(n).

**Review by derrabus — CHANGES_REQUESTED** [review]
Does this change actually solve a problem for you? Did you run any benchmarks?

The obvious blocker for me is the fact that the constructor call gets unnecessarily expensive with your change.

Also since the logic behind the `count()` method gets more complex with your change, we will probably need more tests around the calculation.

That being said, I'm not sure the added complexity is worth it, tbh.

File: .gitignore (hunk 1 above)
Comment (snipershady) [reply .gitignore]: done

File: src/ArrayCollection.php (hunk 2 above)
Comment (snipershady) [reply src/ArrayCollection.php]: only if you create a collection from another collection.
All new collection costs O(1)

File: .gitignore (hunk 1 above)
Comment (greg0ire) [reply .gitignore]: Pro tip: [use a global gitignore](https://greg0ire.fr/git-gud/#/23)

**Comment by snipershady** [comment]
The Static Analysis problem, from PHPstan, wasn't a problem of this PR
I hadn't modified that part of the doc

**Comment by greg0ire** [comment]
> The Static Analysis problem, from PHPstan, wasn't a problem of this PR

Are you willing to bet? :stuck_out_tongue: There's an easy way you could make sure of that, you know :wink:

File: src/ArrayCollection.php [hunk 3]
```diff
@@ -98,7 +104,7 @@ public function first()
      * @psalm-template K of array-key
      * @psalm-template V
      */
-    protected function createFrom(array $elements)
+    protected function createFrom(array $elements): static
```
Comment (greg0ire) [inline src/ArrayCollection.php:107]: Also, the "changes" view makes it obvious which part of your changes is causing the issue.

File: src/ArrayCollection.php (hunk 2 above)
Comment (derrabus) [reply src/ArrayCollection.php]: > only if you create a collection from another collection.

Which is a totally valid thing to do! This is a blocker!

File: src/ArrayCollection.php [hunk 4]
```diff
@@ -65,6 +70,7 @@ class ArrayCollection implements Collection, Selectable, Stringable
     public function __construct(array $elements = [])
     {
         $this->elements = $elements;
+        $this->size     = empty($elements) ? 0 : count($elements);
```
Comment (derrabus) [inline src/ArrayCollection.php:73]: You don't need to avoid `count()` _on an empty array_. That's a cheap operation. `count`ing non-empty arrays preemptively is what troubles me.

File: src/ArrayCollection.php (hunk 3 above)
Comment (derrabus) [inline src/ArrayCollection.php:107]: Adding a native return type is a breaking change. We must not do that in a minor or bugfix release. Apart from that, this change seems unrelated to the goal of this PR.

File: .gitignore [hunk 5]
```diff
@@ -3,4 +3,4 @@
 /build
 /phpcs.xml
 /.phpcs-cache
-/.phpunit.result.cache
+/.phpunit.result.cache
```
Comment (derrabus) [inline .gitignore:6]: Please revert. 🙃

**Review by derrabus — CHANGES_REQUESTED** [review]

**Comment by malarzm** [comment]
FWIW `count` is O(1) when used on arrays, PHP tracks number of elements in a hash table on its own https://github.com/php/php-src/blob/ea6110f52e40c2b2112364df6d6ca99cb81ab63f/Zend/zend_types.h#L407
